### PR TITLE
Fix qasm export for gates with same name

### PR DESCRIPTION
**Pull request by Dpbm**
### Summary
Before this fix, the resulting qasm for a circuit with inner gates sequentially added with the same name, was wrong.
Some Algorithms, like Grover's algorithm, produce a wrong qasm by this problem.  

To fix it, just a name recheck was added!

### Details and comments
See #10162 for further information

### CODE REVIEW

File: test/python/circuit/test_circuit_qasm.py [hunk 1]
```diff
@@ -768,6 +768,39 @@ def test_opaque_output(self):
...
+gate a q0,q1,q2 {{ h q0; h q1; h q2; }}
+gate a_{0} q0,q1,q2 {{ a q0,q1,q2; x q0; x q1; x q2; }}
+qreg q[3];
+a_{0} q[0],q[1],q[2];
+z q[0];
+z q[1];
+z q[2];
+""".format(
+            gate_a_id
+        )
```
Comment (jakelishman) [inline test/python/circuit/test_circuit_qasm.py:801]: Very minor, but it'll be a little easier to read this if you use f-string formatting.  It basically looks identical, except you have the opening quote look like `f"""` (e.g. `expected_output = f"""OPENQASM 2.0;`), in place of ``{0}`` on line 795 you can just write `{gate_a_id}`, and you won't need the `.format()` call.

**Review by jakelishman — COMMENTED** [review]
Thanks for looking at this.  It's a nuisance of an issue, really, because the motivating example is not _strictly_ valid by our data model, but in practical terms we probably ought to support it in the OpenQASM 2 export.

I think the way you've done things here is fine, and solves the problem.  I did idly wonder if it might be "neater" to specify a placeholder gate that we put in `gates_to_define` ahead of the recursion, then move it to the end of the order and give it its true definition after, but it's probably not any better.

I left a minor comment about the test, but other than that we'll be ok to merge this, thanks!

**Comment by Dpbm** [comment]
all done 😁

**Review by jakelishman — APPROVED** [review]
Thanks for the fix, and the quick update!

**Pull request merged**